src/emulator.ts

```diff
@@ -0,0 +1,245 @@
+                        .map(async (emu) => {
+                            if (emu.hash === optionsHash) return emu.name;
+                            await deleteEmulator(emu.name);
+                            return undefined;
+                        })
+                )
+            ).filter((emuName) => !!emuName);
+
+            if (existingEmulators.length > 1) throw Error('Emulator name is not unique. This should never happen.');
+            else if (existingEmulators.length === 1 && existingEmulators[0]) {
+                // TODO: We skip this for now, until we implement proper management of snapshot states, so that the snapshot has the correct capabilities.
+                // The current emulator exists already in the correct config, lets check for a snapshot.
+                // const snapshotList = (await listSnapshots())[existingEmulators[0]];
+                // emulator.resetSnapshotName = snapshotList?.find((s) => s.startsWith('cyanoacrylate-ensured'));
+            } else await createEmulator(emulator.name, emulatorOptions);
+
+            emulator.createdByLib = true;
+        } else if (emulatorRunTargetOptions.emulatorName) {
+            emulator.name = emulatorRunTargetOptions.emulatorName;
+            emulator.resetSnapshotName = emulatorRunTargetOptions.snapshotName;
+        } else throw new Error('Could not start emulator: No emulator config or name.');
+
+        const startEmulatorOptions = emulatorRunTargetOptions.startEmulatorOptions;
+        if (startEmulatorOptions?.headless === true) emulator.startArgs.push('-no-window');
+        if (startEmulatorOptions?.audio !== true) emulator.startArgs.push('-no-audio');
+        if (startEmulatorOptions?.ephemeral !== false) emulator.startArgs.push('-no-snapshot-save');
+        if (startEmulatorOptions?.hardwareAcceleration?.mode)
+            emulator.startArgs.push('-accel', startEmulatorOptions?.hardwareAcceleration?.mode);
+        if (startEmulatorOptions?.hardwareAcceleration?.gpuMode)
+            emulator.startArgs.push('-gpu', startEmulatorOptions?.hardwareAcceleration?.gpuMode);
+
+        return emulator;
+    }
+
+    /**
+     * Start the emulator. If it is already running, this will do nothing, expect if `options.forceRestart` is set.
+     * Then, it kills the running process and starts the emulator again.
+     *
+     * This also creates a new `AbortController` if the current one had been activated.
+     *
+     * @param options
+     *
+     * @returns An `AbortSignal` which is triggered if the emulator crashes.
+     */
+    async start(options?: { forceRestart: boolean }) {
+        if (!this.name) throw new Error('A name is missing. The emulator was not initialized.');
+
+        if (options?.forceRestart && !this.hasExited && this._proc) {
+            // We can do this, since `this.start()` throws away the previous process and its listeners.
+            this._proc.on('exit', () => this.start());
+            return this.kill();
```

Comment:
As far as I understand the docs, this resolves the promise if the signal has been sent, not if the process is actually terminated.

Reply:
You are correct. I was confused because `this.kill()` returned a Promise, but that's only because it's our own async function. `process.kill()` is synchronous and doesn't wait for the process exit:

```sh
❯ node test.mjs
Started program with PID 1915302
PROG: My PID is 1915302
Sending SIGTERM to child process...
Sent SIGTERM to child process...
PROG: SIGTERM received, waiting 10 seconds...
PROG: Exiting now
Child process exited with code 0
```

`test.mjs`:

```js
import { spawn } from 'child_process';

const proc = spawn('node', ['prog.mjs'], { stdio: 'inherit' });

console.log(`Started program with PID ${proc.pid}`);

// Send SIGTERM to the child process after 5 seconds
setTimeout(() => {
    console.log('Sending SIGTERM to child process...');
    proc.kill('SIGTERM');
    console.log('Sent SIGTERM to child process...');
}, 5000);

proc.on('exit', (code) => {
    console.log(`Child process exited with code ${code}`);
});
```

`prog.mjs`:

```js
console.log('PROG: My PID is', process.pid);

process.on('SIGTERM', () => {
    console.log('PROG: SIGTERM received, waiting 10 seconds...');
    setTimeout(() => {
        console.log('PROG: Exiting now');
        process.exit(0);
    }, 10000);
});

// Keep the process running
setInterval(() => {}, 1000);
```